hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/Schema.java

```diff
@@ -0,0 +1,185 @@
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.serialization.impl.compact;
+
+import com.hazelcast.internal.serialization.impl.compact.schema.SchemaDataSerializerHook;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.FieldType;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeMap;
+
+/**
+ * Represents the schema of a class.
+ * Consists of field definitions and the class name.
+ */
+public class Schema implements IdentifiedDataSerializable {
+
+    private String typeName;
+    private TreeMap<String, FieldDescriptor> fieldDefinitionMap;
+    private int numberOfComplexFields;
+    private int primitivesLength;
+    private transient long schemaId;
+    private boolean isSchemaIdSet;
+
+    public Schema() {
+    }
+
+    public Schema(String typeName, TreeMap<String, FieldDescriptor> fieldDefinitionMap,
+                  int numberOfComplexFields, int primitivesLength) {
+        this.typeName = typeName;
+        this.fieldDefinitionMap = fieldDefinitionMap;
+        this.numberOfComplexFields = numberOfComplexFields;
+        this.primitivesLength = primitivesLength;
+    }
+
+    /**
+     * The class name provided when building a schema
+     * In java, when it is not configured explicitly, this falls back to full class name including the path.
+     *
+     * @return name of the class
+     */
+    public String getTypeName() {
+        return typeName;
+    }
+
+    public Collection<FieldDescriptor> getFields() {
+        return fieldDefinitionMap.values();
+    }
+
+    public Set<String> getFieldNames() {
+        return fieldDefinitionMap.keySet();
+    }
+
+    public int getNumberOfVariableLengthFields() {
+        return numberOfComplexFields;
+    }
+
+    public int getPrimitivesLength() {
+        return primitivesLength;
+    }
+
+    public int getFieldCount() {
+        return fieldDefinitionMap.size();
+    }
+
+    public FieldDescriptor getField(String fieldName) {
+        return fieldDefinitionMap.get(fieldName);
+    }
+
+    public boolean hasField(String fieldName) {
+        return fieldDefinitionMap.containsKey(fieldName);
+    }
+
+    public long getSchemaId() {
+        assert isSchemaIdSet;
+        return schemaId;
+    }
+
+    public void setSchemaId(long schemaId) {
+        this.isSchemaIdSet = true;
+        this.schemaId = schemaId;
+    }
+
+    public boolean isSchemaIdSet() {
+        return isSchemaIdSet;
+    }
+
+    @Override
+    public String toString() {
+        return "Schema {"
+                + " className = " + typeName
+                + ", map = " + fieldDefinitionMap
+                + '}';
+    }
+
+    @Override
+    public void writeData(ObjectDataOutput out) throws IOException {
```

Comment:
is there any proof that this generates unique byte sequence considering all possibilities? consider two var sized fields, we write name+typeid+index.

Reply:
`fieldDefinitionMap` is always sorted. So same class should have the same byte sequence always. 
What is your suggestion on this? Put a test to prove that it works ! But on the same codebase that does not make much sense. We may test this across languages when we have them.